### ISSUE TITLE
Another way to get "UserDetailsService is required" when using token refresh

### ISSUE THREAD

**Issue by mgumerov** [spring-projects/spring-security#5901]
### Summary

If UserDetailsService is declared as bean but not explicitly configured as AuthorizationManagerBuilder's details service, attempt to use "refresh_token" grant type results in aforementioned exception. At the same time, "password" grant authentication successfully discovers and uses that details service.

### Actual Behavior

As a prerequisite, we have UserDetailsService as a bean. But we did not explicitly pass it to AuthorizationServerEndpointsConfigurer in any way (userDetailsService or via AuthorizationManagerBuilder.inMemoryAuthentication or something like that), relying in it being picked up with autoconfiguration and/or autowiring.

It actually gets picked up by some ProviderManager and get its way to "global" AuthenticationManager (constructed via DefaultPasswordEncoderAuthenticationManagerBuilder by default, so it's ProviderManager), but it's not reflected as that AuthenticationManager builder's "defaultUserDetailsService". When performing OAuth2 authentication with "password" grant that works well enough, because ProviderManager delegates "authenticate" invocation to its AuthorizationProvider which approves it using its details service. However when performing "refresh_token" grant authentication, Spring Security code (see AuthenticationManagerDelegator.authenticate) does not invoke ProviderManager's authenticate() but instead attempts to query UserDetailsService immediately. And it only respects defaultUserDetailsService, ignoring anything wired into AuthorizationProviders.

Hence, it does not find any AuthorizationManagerBuilder with defaultUserDetailsService, and throws exception.

### Expected Behavior

Not firing exception; somehow discover UserDetailsService for "refresh_token" authentication if that UserDetailsService was able to be discovered for "password" authentication.  Maybe UserDetailsServiceDelegator should be modified to also scan AuthorizationManagerBuilder's AuthorizationProviders not just AuthorizationManagerBuilder's defaultUserDetailsService.

### Configuration

Basically nothing except what's autoconfigured by spring-security-oauth2-autoconfigure 2.1.0.M2 (OAuth2AuthorizationServerConfiguration; yes I used it and it's not technically part of Spring Security, but inability of the latter to discover UserDetailsService is irrelevant to that).
I only added my beans implementing UserDetailsService and PasswordEncoder.

### Version

Spring Security from Spring Boot 2.1.0.M3 (so, Spring 5), but I saw very similar code bundled with Spring Boot 1.5.10.RELEASE (that's Spring 4).

**Comment by mgumerov** [comment]
Actually I was not correct when proposing solution. The whole idea of loading UserDetails via ProviderManager's definition is defective. Generally speaking, there's no telling exactly what provider will be serving user's authentication before it's served, because providers are simply tried one-by-one from first one. For that cause, for UserDetailsServiceDelegator (which is also something used in general case) it's not correct to use some other UserDetails than one in providers, neither it would be correct to just pick any provider which has UserDetails: that might be not the one to actually serve the authentication.

It occurs to me, that the only correct way is to perform authentication (maybe somehow in read-only mode, without modifying application state - though again in general case it's not possible because auth may involve calls, say, to external systems not providing read-only user checks; that auth will serve to determine which of providers is serving the auth, and then UserDetailsServiceDelegator will query that provider's details services to load user details.
Or, as a middleground, UserDetailsServiceDelegator should never be used at all. Instead, say,
PreAuthenticatedAuthenticationProvider (relevant to refresh_token problem) should never try to rely on what user details service is supplied to it by default, and always require explicitly configuring it with one specific UserDetailsService. Of course, once this road is taken, "password" grant should also only be serviced with that one UserDetailsService and not something else, and thus also require specifying it explicilty (it should be one option for both of them, actually).

Of course, that latter part regarding PreAuthenticatedAuthenticationProvider etc. is primarily an issue for spring-security-oauth project, but the whole idea of throwing away UserDetailsServiceDelegator is clearly relevant to basic spring-security.

**Comment by rwinch** [comment]
Thanks for the report. It is a bit confusing, but you are using the older version of Spring Security OAuth which means this issue belongs in https://github.com/spring-projects/spring-security-oauth/issues Can you please create a ticket there?

You can learn more about our OAuth Roadmap at https://spring.io/blog/2018/01/30/next-generation-oauth-2-0-support-with-spring-security

**Comment by mgumerov** [comment]
Did you read the follow-up?

> Of course, that latter part regarding PreAuthenticatedAuthenticationProvider etc. is primarily an issue for spring-security-oauth project, but the whole idea of throwing away UserDetailsServiceDelegator is clearly relevant to basic spring-security.

Hence I felt to place bugreport in both trackers, because there is something to do on both sides. Don't you agree with me that UserDetailsServiceDelegator's logic is initially flawed and it should not be used except probably in very specific conditions? Or do you think issues with UserDetailsServiceDelegator (which is not a part of Spring OAuth) also belong to that tracker and not here?

**Comment by rwinch** [comment]
I think the issue belongs where the problem is and we can go from there to determine what the proper solution is.
